[Clang] Fix dependency computation for pack indexing expression

cor3ntin opened the pull request:
Given `foo...[idx]` if idx is value dependent, the expression is type dependent.

Fixes #91885
Fixes #91884

cor3ntin committed:
[Clang] Fix dependency computation for pack indexing expression

Given `foo...[idx]` if idx is value dependent, the expression
is type dependent.

Fixes #91885
Fixes #91884

zyn0217 commented on clang/lib/AST/ComputeDependence.cpp:
```diff
@@ -376,6 +376,9 @@ ExprDependence clang::computeDependence(PackExpansionExpr *E) {
 ExprDependence clang::computeDependence(PackIndexingExpr *E) {
   ExprDependence D = E->getIndexExpr()->getDependence();
+  if (D & ExprDependence::Value)
+    D |= ExprDependence::TypeInstantiation;
```
Does it mean "if the index expression is value-dependent, then the whole `PackIndexingExpr` is type-dependent despite its expressions"? For example,
```cpp
template <int N, int... E>
auto foo() {
  return E...[N];
}
```
Should `E...[N]` be treated as type-dependent regardless of the result being always an int?

zyn0217 replied:
Probably `E->getPackIdExpression()->getDependence() & ExprDependence::TypeInstantiation`?

cor3ntin replied:
> Should E...[N] be treated as type-dependent regardless of the result being always an int?

Yes. (We don't do anything to look at whether all the expressions would have the same type)

cor3ntin replied:
NVM, you are right, we can simplify this whole thing, I'll do that shortly

cor3ntin committed:
address review feedback

cor3ntin committed:
add additional test

cor3ntin committed:
Fix tests?

cor3ntin committed:
Fix getDecltypeForExpr + cleanup

zyn0217 commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -9806,7 +9806,7 @@ QualType Sema::BuildCountAttributedArrayType(QualType WrappedTy,
 /// [dcl.type.simple]p4 and C++11 [expr.lambda.prim]p18.
 QualType Sema::getDecltypeForExpr(Expr *E) {
-  if (E->isTypeDependent())
+  if (E->isInstantiationDependent())
```
Is it... right? I was assuming this would bring such regressions e.g.

```cpp
template <class T>
const int val = sizeof(T);

template <class T>
void foo() {
  decltype(val<T>);  // <-- this is instantiation-dependent, but not type-dependent.
}
```

Perhaps we can move it to line 9816 where the `PackIndexingExpr` is handled?

cor3ntin replied:
> Perhaps we can move it to line 9816 where the PackIndexingExpr is handled?

I would be slightly concerned about having specific handling of `PackIndexingExpr` all over the place,
or about  `decltype` behaving differently for pack indexing.

Then again, I'm not sure there is a scenario where it matters.

cor3ntin replied:
So here is a thought.
If the pack indexing expression is not type dependent, decltype should just produce the type of that expression.
By that logic we do want to modify decltype so that it does something more clever than to try to use the selected expression.

I'll try to do something in that direction, but it's unclear to me under which set of circumstances we can observe that.

cor3ntin committed:
More correct decltype type computation

If a pack indexing expression has no selected expression
but has a non-tytpe dependent pattern, we can use that type to
compute the decltyoe type.

zyn0217 reviewed (approved):
Thanks for the prompt fix! I think this looks good now, but please wait for @erichkeane for the final stamp.
